includes/forms/functions.php

```diff
@@ -1106,27 +1117,55 @@ function _give_get_prefill_form_field_values( $form_id ) {
 			// Zipcode
 			'card_zip'        => $donor_address['zip'],
 		];
-	endif;
+	}
 
-	// Bailout: Auto fill form field values only form form which donor is donating.
+	// donor info from session
+	$give_purchase_data = Give()->session->get('give_purchase');
 	if (
-		empty( $_GET['form-id'] )
-		|| ! $form_id
-		|| ( $form_id !== absint( $_GET['form-id'] ) )
+		// get data from session only if related to same form
+		isset($give_purchase_data['post_data']['give-form-id'])
+		&& $give_purchase_data['post_data']['give-form-id'] == $form_id
 	) {
-		return $logged_in_donor_info;
-	}
+		// from Give_Donor which related to session give_email
+		if (!empty($give_purchase_data['post_data']['give_email'])) {
+			$donor_email = $give_purchase_data['post_data']['give_email'];
+			$donor = new Give_Donor($donor_email);
+			$donor_address = $donor->get_donor_address();
+			$donor_from_session_info = [
+				'give_first'      => $donor->get_first_name(),
+				'give_last'       => $donor->get_last_name(),
+				'give_title'      => $donor->get_meta( '_give_donor_title_prefix', true ),
+				'company_name'    => $donor->get_company_name(),
+				'give_email'      => $donor_email,
+				'card_address'    => $donor_address['line1'],
+				'card_address_2'  => $donor_address['line2'],
+				'billing_country' => $donor_address['country'],
+				'card_state'      => $donor_address['state'],
+				'card_city'       => $donor_address['city'],
+				'card_zip'        => $donor_address['zip'],
+			];
+		}
 
-	// Get purchase data.
-	$give_purchase_data = Give()->session->get( 'give_purchase' );
+		// from last POST data
+		if(!empty($give_purchase_data['post_data'])) {
+			$session_info = $give_purchase_data['post_data'];
+		}
+	}
 
-	// Get donor info from form data.
-	$give_donor_info_in_session = empty( $give_purchase_data['post_data'] )
-		? []
-		: $give_purchase_data['post_data'];
+	$donor_info = array_merge(
+		$logged_in_user_info,
+		$logged_in_donor_info,
+		$donor_from_session_info,
+		$session_info
+	);
 
-	// Output.
-	return wp_parse_args( $give_donor_info_in_session, $logged_in_donor_info );
+	// allow third party plugins filtering the result
+	return apply_filters(
```

Comment:
I don't mind introducing a new filter here, but I would rather not change the functionality to introduce a new  data source (the session). I can see the feature being helpful, but would rather provide this as a code snippet which uses the new filter, which would allow site owners to opt-in to this change of behavior.